File: assiist_front_end/lib/core/repositories/task_repository.dart
import 'package:assiist_front_end/core/models/task.dart';

/// Abstract interface for task data operations.
abstract class TaskRepository {
  /// Retrieves all tasks associated with a specific contact ID.
  Future<List<Task>> getTasksForContact(String contactId);

  /// Retrieves all tasks relevant to the current user (e.g., for the dashboard).
  /// Might involve filtering by status, due date, etc., depending on implementation.
  Future<List<Task>> getAllTasksForUser();

  /// Creates a new task directly.
  Future<Task> createTask(Task task);

  /// Updates an existing task.
  Future<Task?> updateTask(
    String taskId,
    String contactId, // Needed to locate the task in subcollection
    Map<String, dynamic> updates,
  );

  /// Deletes a task.
  Future<bool> deleteTask(
    String taskId,
    String contactId, // Needed to locate the task in subcollection
  );

  /// Retrieves a task by its ID and contact ID.
  Future<Task?> getById(String taskId, String contactId);

  /// Retrieves a task by ID without requiring contactId (direct lookup for performance)
  Future<Task?> getTaskById(String taskId);
}

File: assiist_back_end/db/repositories/interfaces/task_repository.py
from abc import ABC, abstractmethod
from typing import List, Optional, Dict, Any, Tuple
import uuid

from assiist_back_end.models.task import Task
from assiist_back_end.models.revision import RevisionHistory
from assiist_back_end.db.repositories.interfaces.revision_repository import RevisionHistoryRepository

class TaskRepository(ABC):
    """Interface for data operations related to Tasks."""

    @abstractmethod
    async def add(self, user_id: str, contact_id: str, task: Task) -> Task:
        """Adds a new task to a contact's subcollection."""
        pass

    @abstractmethod
    async def get_by_id(self, user_id: str, contact_id: str, task_id: str) -> Optional[Task]:
        """Gets a specific task by its ID."""
        pass

    @abstractmethod
    async def get_for_contact(self, user_id: str, contact_id: str) -> List[Task]:
        """Gets all tasks for a specific contact."""
        pass

    @abstractmethod
    async def update(self, user_id: str, contact_id: str, task_id: str, update_data: Dict[str, Any]) -> Optional[Task]:
        """Updates an existing task."""
        pass

    @abstractmethod
    async def delete(self, user_id: str, contact_id: str, task_id: str) -> bool:
        """Deletes a specific task, ensuring it belongs to the contact/user."""
        pass

    @abstractmethod
    async def get_all_for_user(self, user_id: str) -> List[Task]:
        """Retrieves all tasks associated with a specific user across all contacts."""
        pass

    @abstractmethod
    async def add_with_revision_history(
        self, 
        task: Task, 
        original_message: str,
        context: Dict[str, Any],
        revision_repo: RevisionHistoryRepository
    ) -> Tuple[Task, RevisionHistory]:
        """Creates a task along with its revision history atomically."""
        pass

    # Potentially add methods for querying tasks by status, assignee, etc. 